pkg/sink/sink.go

```diff
@@ -136,3 +155,64 @@ func createRequestURI(apiVersion, namePlural, namespace string) string {
 	uri = path.Join(uri, namePlural)
 	return uri
 }
+
+func (r Resource) secureEndpoint(trigger triggersv1.Trigger, headers http.Header, payload []byte) error {
+
+	params := []pipelinev1.Param{}
+
+	params = append(params, trigger.TriggerValidate.Params...)
+	params = append(params, pipelinev1.Param{
+		Name: "Payload",
+		Value: pipelinev1.ArrayOrString{
+			Type:      pipelinev1.ParamTypeArray,
+			StringVal: string(payload),
+		},
+	})
+
+	for key := range headers {
+		params = append(params, pipelinev1.Param{
+			Name: key,
+			Value: pipelinev1.ArrayOrString{
+				Type:      pipelinev1.ParamTypeArray,
+				StringVal: headers.Get(key),
+			},
+		})
+	}
+
+	tr := &pipelinev1.TaskRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    r.EventListenerNamespace,
+			GenerateName: trigger.TriggerValidate.TaskRef.Name,
+		},
+		Spec: pipelinev1.TaskRunSpec{
+			Inputs: pipelinev1.TaskRunInputs{
+				Params: params,
+			},
+			TaskRef:        trigger.TriggerValidate.TaskRef,
+			ServiceAccount: trigger.TriggerValidate.ServiceAccount,
+		},
+	}
+
+	tr, err := r.PipelineClient.TektonV1alpha1().TaskRuns(r.EventListenerNamespace).Create(tr)
+	if err != nil {
+		return err
+	}
+
+	for {
+		tr, err := r.PipelineClient.TektonV1alpha1().TaskRuns(r.EventListenerNamespace).Get(tr.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if tr.IsSuccessful() {
+			break
+		}
+
+		time.Sleep(validateTaskWait)
```

Comment:
Hmmm, this seems a bit brittle. We are calling `secureEndpoint` for each trigger in a event listener spec in a for loop and then here we loop forever until the taskrun is done. 
So, if one of these validation taskruns takes a long time, other triggers will not be processed. Also, we are completely dependent on the TaskRun to behave properly (i.e. it either finishes successfully or times out properly). 

Some ideas:

* We process each trigger in its own goroutine so that one validation taskrun being slow does not block the others.

*  Instead of a for loop that goes on for ever, we should have some number of retries or give up after a timeout

* Add number of retries and a reasonable timeout value to these taskruns 

* In the future, instead of polling at fixed intervals, we could consider a informer/reconciliation based approach?

Reply:
@dibyom What do you think of using a watch? For now, I have done using retries and taskRunPollingInterval.

Reply:
Oh. And I created a function to execute trigger which runs under a goroutine.